[SystemC] Add SCModuleOp and AliasOp

maerhart opened the pull request:
Includes the `systemc.module` and `systemc.alias` operation definitions as well as a basic lowering from HW. The lowering will be extended in another PR once other necessary operations such as `systemc.func` and `systemc.method` are added.

seldridge commented on include/circt/Dialect/SystemC/SystemCStructure.td:
```diff
@@ -11,8 +11,72 @@
+
+  let extraClassDeclaration = [{
+    // Implement RegionKindInterface.
+    static RegionKind getRegionKind(unsigned index) { return RegionKind::Graph;}
```
This lack of a space after`;` bothers me, but I also grok tha tthis is to keep it under 80 chars. Consider:
```suggestion
    static RegionKind getRegionKind(unsigned index) { 
      return RegionKind::Graph;
    }
```

maerhart replied:
You might also want to update it in `HWStructure.td` then 😉

seldridge commented on lib/Conversion/HWToSystemC/HWToSystemC.cpp:
```diff
@@ -21,6 +22,112 @@
+        directions.push_back(systemc::PortDirection::Input);
+      else if (port.isOutput())
+        directions.push_back(systemc::PortDirection::Output);
+      else
+        directions.push_back(systemc::PortDirection::InOut);
```
Is it possible to ever exercise this code via a HW -> SystemC module conversion? Should this be an error for now?

maerhart replied:
You can, for example, exercise it with `hw.module @someModule (%inoutArg: !hw.inout<i32>) -> () {}` which is converted to `systemc.module @someModule (sc_inout %inoutArg: !hw.inout<i32>) {}`. Should have added that to a test. We might also have to peel off the `!hw.inout` wrapper type. But I think these inout values are only really used by the SV dialect, so to error out can also be a strategy for now.

seldridge commented on lib/Conversion/HWToSystemC/HWToSystemC.cpp:
```diff
@@ -21,6 +22,112 @@
+    // SystemC module port types are all expressed as block arguments to the op,
+    // so collect all of the types and add them as arguments to the SystemC
+    // module.
+    SmallVector<Type, 4> portTypes(moduleInputs);
```
Nit: `moduleInputs` is single-use and could be sunk into the one use.

seldridge commented on lib/Conversion/HWToSystemC/HWToSystemC.cpp:
```diff
@@ -21,6 +22,112 @@
+    portNames.append(args.begin(), args.end());
+    portNames.append(results.begin(), results.end());
+
+    // Create the SystemC module. Note that no parameterized modules are
+    // supported yet.
```
Is there a check that the module is not parameteric anywhere?

seldridge commented on lib/Conversion/HWToSystemC/HWToSystemC.cpp:
```diff
@@ -21,6 +22,112 @@
+                                                  rewriter.getUnknownLoc()));
+    });
+
+    // Erase the HW module.
+    rewriter.eraseOp(module);
```
Is it safe to erase the HW module at this point given that it may have uses (instantiations) in another HW module?

maerhart replied:
If you lower one isolated design in its entirety, it is save (will error out on the instance not being lowerable). When implementing the instance lowering we have to make sure that it is aware of the referred module op already being lowered. If we only lower one (or a subset) of module to systemc, or there is an instantiation in another design refering to a module in this one via an external module, then this implementation is not save. But partially lowering a design is not supported yet and for the other case we need to implement some interop, which I would leave for a later PR. So I would suggest we erase if for now and don't consider such external references for now so we don't have to deal with name uniquification, etc. in this PR.

seldridge commented on lib/Dialect/SystemC/SystemCAttributes.cpp:
```diff
@@ -0,0 +1,108 @@
+
+PortDirectionsAttr PortDirectionsAttr::get(MLIRContext *context,
+                                           ArrayRef<PortDirection> directions) {
+  APInt storage(2 * directions.size(), 0);
+  for (size_t i = 0; i < directions.size(); ++i) {
+    storage.insertBits(APInt(2, static_cast<uint32_t>(directions[i])), 2 * i);
+  }
```
This storage is clever. 👍

maerhart replied:
Thanks!

seldridge reviewed (commented):
This is looking great. Some questions and nitty comments throughout.

maerhart commented on include/circt/Dialect/SystemC/SystemCAttributesImpl.td:
```diff
@@ -0,0 +1,73 @@
+
+def PortDirectionsAttr : SystemCAttrDef<"PortDirections"> {
+  let summary = "Port directions attribute";
+  let description = [{
+  }];
```
Ah right, forgot to populate this empty description 🤦

maerhart commented:
I think I addressed all the comments, so I'll merge it to unblock another PR. If there is anything else, I'm happy to address it in a follow-up commit.
